A80-pid.md

```diff
@@ -0,0 +1,284 @@
+
+  // Minimum allowable weight. Weights proposed by the PID controller falling below this value will be capped.
+  // This prevents weights from dropping to zero, which could occur if only a subset of clients uses PID and decreasing weights
+  // no longer effectively corrects the imbalance. Default is 0.1.
+  google.protobuf.FloatValue min_weight = 6;
+}
+```
+
+### PID controller
+
+A PID controller is a control loop feedback mechanism that continuously calculates an error value as the difference between a desired setpoint (`referenceSignal`) and a measured process variable (`actualSignal`). It then applies a correction based on proportional, integral, and derivative terms (denoted P, I, and D respectively), hence the name.
+
+In our implementation, we will not be using the integral part. The integral component is useful for speeding up convergence when the `referenceSignal` changes sharply. In our case, we will be converging the load on the subchannels to a mean value, which is mostly stable.
+
+Here is a sample implementation in pseudo-code:
+
+```
+pidController class {
+  proportionalGain float
+  derivativeGain float
+
+  controlError float
+
+  update(referenceSignal float, actualSignal float, samplingInterval duration) float {
+    previousError = this.controlError
+    // Save last controlError so we can use it to calculate derivative during next update
+    this.controlError = referenceSignal - actualSignal
+    controlErrorDerivative = (this.controlError - previousError) / samplingInterval.Seconds()
+    controlSignal = this.proportionalGain * this.controlError +
+                    this.derivativeGain * controlErrorDerivative
+    return controlSignal
+  }
+}
+```
+
+The `update` method is expected to be called on a regular basis, with `samplingInterval` being the duration since the last update. The return value is the control signal which, if applied to the system, should minimize the control error. In the next section, we'll discuss how this control signal is converted to `wrr` weight.
+
+The `proportionalGain` and `derivativeGain` parameters are taken from the LB config. `proportionalGain` should be scaled by the `WeightUpdatePeriod` value. This is necessary because `controlErrorDerivative` is inversely proportional to the sampling interval, which in turn is close to the `WeightUpdatePeriod` as we will be updating the PID state once per `WeightUpdatePeriod`. If `WeightUpdatePeriod` is too small, `controlErrorDerivative` becomes too large and dominates the resulting control error.
+
+### Extending WRR balancer
+
+The `pid` balancer reuses 90% of the wrr code. The proposal is to refactor the `wrr` codebase and introduce several hooks that allow other balancers, like `pid`, to reuse the code efficiently without the need for duplication. Initially, these hooks will remain internal to avoid prematurely establishing a new public API. This approach is mostly language-specific, but the general plan is as follows:
```

Comment:
I could see things going two ways on this:
- make wrr pluggable and public
- integrate PID upstream.
In a way with the precedent of Spotify copy pasting WRR for their own need I think we have a case for opening up the WRR interface, as there are 3 known use cases.